File: levelledmobs-plugin-bukkit/src/main/java/io/github/arcaneplugins/levelledmobs/bukkit/util/math/TimeUtils.java
package io.github.arcaneplugins.levelledmobs.bukkit.util.math;

import java.util.function.Function;
import javax.annotation.Nonnull;

public class TimeUtils {

    private TimeUtils() throws IllegalAccessException {
        throw new IllegalAccessException("Attempted instantiation of utility class");
    }

    public static long parseTimeToTicks(
        final @Nonnull Object o
    ) {
        if(o instanceof final String s) {
            try {
                return Long.parseLong(s);
            } catch(NumberFormatException ignored) {}

            final Function<Integer, Long> parser = (suffixLen) ->
                Long.parseLong(s.substring(0, s.length() - suffixLen));

            if(s.endsWith("t")) {
                return parser.apply(1);
            } else if(s.endsWith("s")) {
                return parser.apply(1) * 20;
            } else if(s.endsWith("ms")) {
                return (long) (parser.apply(2) * 0.002d);
            } else if(s.endsWith("m")) {
                return parser.apply(1) * 1_200;
            } else if(s.endsWith("h")) {
                return parser.apply(1) * 72_000;
            } else if(s.endsWith("d")) {
                return parser.apply(1) * 1_728_000;
            } else {
                throw new IllegalArgumentException("Invalid time unit '%s'."
                    .formatted(s.substring(s.length() - 1)));
            }
        } else if(o instanceof final Number n) {
            return (long) n;
        } else {
            throw new IllegalArgumentException("Unable to parse type " +
                o.getClass().getSimpleName());
        }
    }

}

File: levelledmobs-plugin-bukkit/src/main/kotlin/io/github/arcaneplugins/levelledmobs/bukkit/util/math/TimeUtils.kt
package io.github.arcaneplugins.levelledmobs.bukkit.util.math

import java.util.function.Function

object TimeUtils {
    fun parseTimeToTicks(
        o: Any
    ): Long {
        return if (o is String) {
            try {
                // default unit: ticks
                return o.toLong()
            } catch (ignored: NumberFormatException) {}
            val parser =
                Function { suffixLen: Int -> o.substring(0, o.length - suffixLen).toLong() }
            if (o.endsWith("t")) {
                parser.apply(1)
            } else if (o.endsWith("s")) {
                parser.apply(1) * 20
            } else if (o.endsWith("ms")) {
                (parser.apply(2) * 0.002).toLong()
            } else if (o.endsWith("m")) {
                parser.apply(1) * 1200
            } else if (o.endsWith("h")) {
                parser.apply(1) * 72000
            } else if (o.endsWith("d")) {
                parser.apply(1) * 1728000
            } else {
                throw IllegalArgumentException(
                    "Invalid time unit '${o.substring(o.length - 1)}'."
                )
            }
        } else (o as? Number)?.toLong()
            ?: throw IllegalArgumentException(
                "Unable to parse type " + o.javaClass.getSimpleName()
            )
    }
}